### PR TITLE
Grant readwrite to root snippets for all users

### LINKED ISSUE #21581: Normal users can't create first snippet in Enterprise version
**Describe the bug**
On a brand new Enterprise install, users don't have the ability to create snippets by default.
In addition, in order to grant access, an admin has to first create a snippet, then change the root permissions.

**Logs**
NA

**To Reproduce**
1) Set up a new instance on Enterprise
2) create a normal user
3) logged in as the normal user, try to create the first snippet, see the error

**Expected behavior**
Unless there's an explicit permission set by an administrator, normal users should be able to create a snippet

**Screenshots**
image

**Information about your Metabase Installation:**
- Your browser and the version: Brave latest
- Your operating system: Ubuntu 20.04.4
- Your databases: Postgres 14
- Metabase version: 1.42.3
- Metabase hosting environment: Docker
- Metabase internal database: Postgres 14

**Severity**
Annoying

**Additional context**
NA

### PR DESCRIPTION
**Pull request by qnkhuat**
Fix #21581

Similar to granting readwrite perms to Root Collection for All Users, we need to grant it for snippets namespace too

### CODE REVIEW

**Comment by qnkhuat** [comment]
Can you help with the `shouldn't update root permissions when changing permissions on a created folder (metabase#17268)` cypress test? @nemanjaglumac

The check permission fails because this PR grants "write" Perms to root snippets for All Users and it's expecting "none".
I think one way we could fix is to change permission on a different Group (not All Users) in that test.
But seems like it's lots of steps to do? Is it safe for us to remove that reproduce test tho?

**Comment by nemanjaglumac** [comment]
> Can you help with the `shouldn't update root permissions when changing permissions on a created folder (metabase#17268)` cypress test? @nemanjaglumac

@qnkhuat Done. Sorry it took me a while to respond. I didn't see this message until today :(
 
> The check permission fails because this PR grants "write" Perms to root snippets for All Users and it's expecting "none". I think one way we could fix is to change permission on a different Group (not All Users) in that test. But seems like it's lots of steps to do? Is it safe for us to remove that reproduce test tho?

We should most definitely not remove the repro.

The solution was fairly simple. I've updated and made this repro more robust. Left some comments along the way. Please let me know if you have any additional questions.

**Review by flamber — CHANGES_REQUESTED** [review]
Blocking this, since this causes changes to existing instance permissions, which might be unintended.
It should just change the default on new instances.

Would probably need an update of the docs too:
https://www.metabase.com/docs/latest/enterprise-guide/sql-snippets.html#permissions

And it's too late for 43 - if implemented, it should be targeting master.

**Comment by camsaul** [comment]
Removing review request for @metabase/backend-developers for now -- please re-request when you've addressed Flamber's feedback

File: resources/migrations/000_migrations.yaml
Comment (qnkhuat) [inline resources/migrations/000_migrations.yaml:11523]: This whole case can be simplified with

```sql
SELECT CASE
          WHEN (user_count > 0) AND (token_count) > 0 THEN 1
          ELSE 0
       END AS RESULT
```

But I prefer to have things explicit here

**Comment by qnkhuat** [comment]
I've updated the migration script to only run for:
- Any new instances
- Existing OSS instances

**Review by noahmoss — APPROVED** [review]
👍 Updated SQL migration seems to work as expected but @flamber should verify as well

**Comment by flamber** [comment]
Is it not possible to implement this without creating migrations? Also, since this is not going into 43, it's perhaps better to get new migration number (if migrations are really required).
I don't know what the intention is, so the product team needs to comment if this is the intended behavior.
And then the documentation needs to be updated to reflect that.

File: resources/migrations/000_migrations.yaml
Comment (luizarakaki) [inline resources/migrations/000_migrations.yaml:11516]: minor: there is a typo here: `instances`

File: resources/migrations/000_migrations.yaml
Comment (luizarakaki) [inline resources/migrations/000_migrations.yaml:11508]: we need to update to 44

**Review by luizarakaki — APPROVED** [review]
I don't see a reason why an admin would need to create a Snippet first and then give access to users. It is sensible to make this feature available out of the box.

So for me, this PR is fine and closes the issue.

**Comment by qnkhuat** [comment]
@flamber I don't see any other way to do this without a migration script. Also, I've managed to run the migration only on new instances or instances that upgrade from OSS -> ee, so I think it should be fine.

Lmk if you still have any concerns with this, otherwise please help me approve. Somehow I can't merge without your approval.

File: resources/migrations/000_migrations.yaml
Comment (flamber) [inline resources/migrations/000_migrations.yaml:11861]: Wouldn't this be enough, since it's not possible to have 0 users, while having a token setting? Also, many setups uses envvars to define token, so there won't be a setting in the appdb.
It would make it easier to read the migration - less is more.

**Review by flamber — APPROVED** [review]

File: resources/migrations/000_migrations.yaml
Comment (qnkhuat) [reply resources/migrations/000_migrations.yaml:11861]: > Also, many setups use envvars to define tokens, so there won't be a setting in the appdb.

My original thinking is that we should run this migration for existing OSS instances so that if they update to EE, this change will be included. 

But if we can't know for sure if an instance is OSS or EE by just reading the DB, I'll make this migration run on new instances only.

**Comment by qnkhuat** [comment]
Update: As Flambed pointed out, there are instances that use envvars to set up token, this means we can't know if an instance is OSS or EE from reading database only.

So I've updated the migration script to run for new instances only -- existing OSS instances won't run this migration.

**Pull request merged**
